src/v/kafka/client/direct_consumer/direct_consumer.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+#include "base/seastarx.h"
+#include "container/chunked_hash_map.h"
+#include "kafka/client/cluster.h"
+#include "kafka/client/direct_consumer/api_types.h"
+
+#include <seastar/core/queue.hh>
+#include <seastar/core/rwlock.hh>
+
+namespace kafka {
+struct metadata_response_data;
+namespace client {
+
+class fetcher;
+class data_queue;
+/**
+ * Simple direct Kafka consumer. It allows user to subscribe to a list of
+ * topic partitions. The consumer maintains broker fetch session, it reacts for
+ * the leadership changes and metadata updates.
+ * Fetch responses from all the brokers are exposed through `fetch_next()`.
+ * The fetch_next() returns whenever any ready to be consumer are available.
+ *
+ * Implementation details:
+ * NOTE:
+ * Fetcher and data queue classes should not be used directly, they are
+ * implementation details of the direct consumer, they were placed into separate
+ * files for readability.
+ *
+ * Direct consumer uses fetcher class to fetch data from the brokers. The
+ * fetcher is responsible for fetching data from a single broker. It maintains a
+ * list of partitions and corresponding fetch offsets. The fetcher loop
+ * constantly queries the broker for data and updates the fetch offsets based on
+ * the received data. Fetcher is also responsible for fetching offsets to apply
+ * the reset policy. Fetcher puts the fetched data into the parent consumer's
+ * data queue. The queue is then accessed through the `fetch_next()`
+ * method.
+ *
+ * The direct consumer maintains a list of partition subscriptions.
+ * When the subscription list is changed the consumer updates the
+ * fetchers state. In order to maintain partition batches ordering the invariant
+ * is held that one partition is never assigned to multiple fetchers at the same
+ * time.
+ *
+ * Fetchers are 'blocked' from fetching while the consumer is updating their
+ * assigned partitions. This is done by acquiring a write lock on the shared
+ * state lock that is being held by the fetchers in the fetch loop.
+ *
+ */
+class direct_consumer {
+public:
+    struct configuration {
+        int32_t min_bytes{1};
+        int32_t max_fetch_size{512_KiB};
+        int32_t partition_max_bytes{128_KiB};
+        offset_reset_policy reset_policy = offset_reset_policy::earliest;
+        std::chrono::milliseconds max_wait_time{100};
+        model::isolation_level isolation_level
+          = model::isolation_level::read_uncommitted;
+    };
+
+    direct_consumer(cluster& cluster, configuration cfg);
+
+    ~direct_consumer();
+    /**
+     * Stops consumer
+     */
+    ss::future<> stop();
```

Comment:
I'm a bit confused when I see `stop` without a `start`. Do I have to call this? When do I need to call this? Is a consumer automatically started?

Reply:
i will add `start()` counterpart